src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java

```diff
@@ -0,0 +1,181 @@
+package com.epam.esm.dao.impl;
+
+
+
+import com.epam.esm.dao.CertificateDao;
+import com.epam.esm.dao.DaoException;
+import com.epam.esm.model.Certificate;
+import com.epam.esm.model.Tag;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataAccessException;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.jdbc.core.BeanPropertyRowMapper;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.PreparedStatementCreator;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.BeanPropertySqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.MultiValueMap;
+
+import java.sql.*;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.epam.esm.dao.RowMapProvider.*;
+import static com.epam.esm.dao.SqlProvider.*;
+
+@Repository
+@Transactional(readOnly = true)
+public class CertificateDaoImpl implements CertificateDao, RowMapper<Certificate> {
+
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+
+    @Autowired
+    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+    private List<Long> getCertificateIdsByTagId(Long tagId){
+        return jdbcTemplate.query(GET_CERTIFICATES_BY_TAG_ID, new Object[] {tagId},
+                (rs, rowNum) -> rs.getLong(1));
+    }
+
+
+    @Override
+    public List<Certificate> getAll() {
+        return jdbcTemplate.query(GET_ALL_CERTIFICATES, new CertificateDaoImpl());
+
+    }
+
+    @Override
+    public List<Certificate> getByParameters(MultiValueMap<String, String> params) throws DaoException {
+        if (CollectionUtils.isEmpty(params))
+            throw new DaoException();
+        final StringBuilder stringBuilder = new StringBuilder("SELECT * FROM gift_certificate WHERE ");
+        if (Optional.ofNullable(params.get("name")).isPresent() && Optional.ofNullable(params.get("description")).isPresent()) {
```

Comment:
Please move logic to separate class.

Reply:
Does it make sence to create a separate class for a single method?

Reply:
You can write your application in one class. Why did you split it into several classes?
The data access layer is only responsible for communicating with the used data store.
https://www.petrikainulainen.net/software-development/design/understanding-spring-web-application-architecture-the-classic-way/

Also please read more about OOD, SOLID, YAGNI, DRY.